// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package research.simulation;

import "swirl_lm/boundary_condition/boundary_conditions.proto";
import "swirl_lm/boundary_condition/boundary_models.proto";
import "swirl_lm/boundary_condition/rayleigh_damping_layer.proto";
import "swirl_lm/equations/pressure.proto";
import "swirl_lm/equations/scalars.proto";
import "swirl_lm/numerics/numerics.proto";
import "swirl_lm/physics/atmosphere/microphysics.proto";
import "swirl_lm/physics/combustion/combustion.proto";
import "swirl_lm/physics/lpt/lpt.proto";
import "swirl_lm/physics/radiation/config/radiative_transfer.proto";
import "swirl_lm/physics/thermodynamics/thermodynamics.proto";
import "swirl_lm/physics/turbulent_combustion/turbulent_combustion.proto";
import "swirl_lm/utility/grid_parametrization.proto";
import "swirl_lm/utility/monitor.proto";
import "swirl_lm/utility/probe.proto";

// Methods for how the simulation time, including the number of cycles
// (`num_cycles`) and number of steps per cycles (`num_steps`), are determined.
// Next id: 4
message SimulationTimeInfo {
  // Parameters are passed in from flags.
  message FromFlags {}

  // Parameters are passed in from the config file.
  message FromConfig {
    optional int32 num_cycles = 1;
    optional int32 num_steps = 2;
  }

  // Parameters are derived from the flow through times that is computed as
  // l / bc_u in the `mean_flow_dim`. Note that this option is only valid if
  // the face with index `inflow_face` along the `mean_flow_dim` is associated
  // with an inflow boundary type.
  // Next id: 5
  message FromFlowThroughTime {
    // The number of cycles the simulation runs. This parameter determines the
    // total number of restart files to be dumped.
    optional int32 num_cycles = 1;

    // The dimension along with the flow through time is computed. Must be one
    // of 0, 1, and 2.
    optional int32 mean_flow_dim = 2;

    // The face from which the inflow is going into the domain. 0 for the lower
    // index end, and 1 for the higher index end.
    optional int32 inflow_face = 3;

    // The total number of flow through times that the simulation needs to
    // cover.
    optional float n_flow_through_time = 4;
  }

  oneof simulation_time_method {
    FromFlags from_flags = 1;
    FromConfig from_config = 2;
    FromFlowThroughTime from_flow_through_time = 3;
  }
}

// Options for performing pre- and post-process in the simulation.
// Next id: 4
message PrePostProcessInfo {
  // Initializes pre- and post-process options from command line flags.
  message FromFlags {}

  // Initializes pre- and post-process options from the config file.
  // Next id: 7
  message FromConfig {
    optional bool apply_preprocess = 1;
    optional int32 preprocess_step_id = 2;
    optional bool preprocess_periodic = 3;
    optional bool apply_postprocess = 4;
    optional int32 postprocess_step_id = 5;
    optional bool postprocess_periodic = 6;
  }

  // Initializes pre- and post-process options from flow through time. Note that
  // this option is only valid if the face with index `inflow_face along the
  // `mean_flow_dim` is associated with an inflow boundary type. The pre- and
  // post-process step id will be rounded to the closest integer multiple of
  // `num_steps`.
  // Next id: 9
  message FromFlowThroughTime {
    optional bool apply_preprocess = 1;
    optional bool preprocess_periodic = 2;
    optional bool apply_postprocess = 3;
    optional bool postprocess_periodic = 4;

    // The dimension along with the flow through time is computed. Must be one
    // of 0, 1, and 2.
    optional int32 mean_flow_dim = 5;

    // The face from which the inflow is going into the domain. 0 for the lower
    // index end, and 1 for the higher index end.
    optional int32 inflow_face = 6;

    // The flow through time at which pre-process is applied.
    optional float preprocess_flow_through_time = 7;

    // The flow through time at which post-process is applied.
    optional float postprocess_flow_through_time = 8;
  }

  oneof pre_post_process_option {
    FromFlags from_flags = 1;
    FromConfig from_config = 2;
    FromFlowThroughTime from_flow_through_time = 3;
  }
}

// Stores information about a sub-grid scale model.
// Next id: 9
message SubGridScaleModel {
  // Parameters required by the constant Smagorinsky model.
  message SmagorinskyModel {
    // The Smagonrinsky constant.
    optional float c_s = 1 [default = 0.18];
    // The turbulent Prandtl number.
    optional float pr_t = 2 [default = 0.3];
    // An indicator of whether the turbulent diffusivity is computed using the
    // turbulent Prandtl number, or from the scalar gradient directly. Default
    // to use the turbulent Prandtl number to match the original behavior.
    optional bool use_pr_t = 3 [default = true];

    enum DeltaFormula {
      UNKNOWN = 0;
      // Delta coefficient is sqrt(dx^2 + dy^2 + dz^2).
      DIAGONAL = 1;
      // Delta coefficient is (dx * dy * dz) ^ (1/3).
      GEOMETRIC_MEAN = 2;
    }

    // Note that the default is DIAGONAL for backward compatibility but
    // GEOMETRIC_MEAN is preferred especially with high aspect ratio grids.
    optional DeltaFormula delta_formula = 4 [default = DIAGONAL];
  }

  // Parameters required by the dynamic Smagorinsky model.
  message DynamicSmagorinskyModel {
    // An indicator of periodicity in the x direction.
    optional bool periodic_x = 1 [default = true];
    // An indicator of periodicity in the y direction.
    optional bool periodic_y = 2 [default = true];
    // An indicator of periodicity in the z direction.
    optional bool periodic_z = 3 [default = true];
  }

  // Parameters required by the Smagorinsky-Lilly model. Default values are
  // retrieved from reference: Stevens, Bjorn, Chin-Hoh Moeng, Andrew S.
  // Ackerman, Christopher S. Bretherton, Andreas Chlond, Stephan de Roode,
  // James Edwards, et al. 2005. “Evaluation of Large-Eddy Simulations via
  // Observations of Nocturnal Marine Stratocumulus.” Monthly Weather Review 133
  // (6): 1443–62.
  message SmagorinskyLillyModel {
    // The Smagonrinsky constant.
    optional float c_s = 1 [default = 0.18];
    // The turbulent Prandtl number.
    optional float pr_t = 2 [default = 0.4];
    // An indicator of whether the turbulent diffusivity is computed using the
    // turbulent Prandtl number, or from the scalar gradient directly. Default
    // to use the turbulent Prandtl number to match the original behavior.
    optional bool use_pr_t = 3 [default = true];
  }

  // Parameters required by the Vreman model.
  // Reference:
  // Vreman, A. W. 2004. “An Eddy-Viscosity Subgrid-Scale Model for Turbulent
  // Shear Flow: Algebraic Theory and Applications.” Physics of Fluids
  // 16 (10): 3670–81.
  message VremanModel {
    // The Smagonrinsky constant. Default value is taken from the reference.
    optional float c_s = 1 [default = 0.18];
    // The turbulent Prandtl number. Default value is taken from the reference.
    optional float pr_t = 2 [default = 0.4];
  }

  oneof sgs_model_type {
    SmagorinskyModel smagorinsky = 1;
    DynamicSmagorinskyModel dynamic_smagorinsky = 2;
    SmagorinskyLillyModel smagorinsky_lilly = 3;
    VremanModel vreman = 4;
  }

  // An upper bound of the turbulent viscosity.
  optional float nu_t_max = 8;

  // An upper bound of the turbulent diffusivity.
  optional float diff_t_max = 5;

  // A lower bound of the turbulent viscosity.
  optional float nu_t_min = 6 [default = 0.0];

  // A lower bound of the turbulent diffusivity.
  optional float diff_t_min = 7 [default = 0.0];
}

// Stores parameters for the microphysics model.
message Microphysics {
  // An option for including the effects of precipitation in the equations of
  // humidity and possibly the energy variable.
  optional bool include_precipitation = 1;
  // An option for including the effects of condensation in the equations for
  // humidity. Note that this option is only valid when using q_c, q_v, and
  // theta as the thermodynamic variables.
  optional bool include_condensation = 2;
  // An option for including sedimentation in the convective flux of the
  // humidity variable.
  optional bool include_sedimentation = 3;
  // Specifies the option of microphysics model.
  oneof model_type {
    Kessler kessler = 4;
    OneMoment one_moment = 5;
  }
}

// Stores the parameters required by an incompressible Navier-Stokes simulation.
// Next id: 43
message SwirlLMParameters {
  // Deprecated/Removed fields.
  reserved 4;

  // Defines the type of the simulation procedure to be used in the solver.
  enum SolverProcedureType {
    // The default option where no solver procedure is specified. Results in an
    // error in this case.
    UNKNOWN = 0;
    // The sequential procedure that first updates pressure then velocity.
    SEQUENTIAL = 1;
    // The variable-density low Mach number solver.
    VARIABLE_DENSITY = 3;

    // Deprecated/Removed fields.
    reserved 2;
  }

  // Defines the method types for density update. This will be deprecated and
  // replaced by `thermodynamics.Thermodynamics`.
  enum DensityUpdateOption {
    // The default option where density is not updated. A constant density is
    // used.
    DENSITY_UPDATE_UNKNOWN = 0;
    // The option for updating density with a linear mixing rule.
    DENSITY_UPDATE_LINEAR_MIXING = 1;
    // The option for updating density from the equation of state.
    DENSITY_UPDATE_EOS = 2;
    // The option for updating density with the water thermodynamics.
    DENSITY_UPDATE_WATER = 3;
  }

  // Defines the type of kernel operator to be used in the simulation algorithm.
  enum KernelOpType {
    // An unimplemented kernel operator type.
    KERNEL_OP_UNKNOWN = 0;
    // The option for using convolution to perform finite difference operations.
    KERNEL_OP_CONV = 1;
    // The option for using slicing to perform finite difference operations.
    KERNEL_OP_SLICE = 2;
    // The option for using matrix multiplication to perform finite difference
    // operations.
    KERNEL_OP_MATMUL = 3;
  }

  // The mesh information.
  optional swirl_lm.GridParametrization grid_params = 1;
  // The type of kernel operator to be used to represent finite difference
  // operators. Default set to `KERNEL_OP_CONV` to match previous behavior.
  optional KernelOpType kernel_op_type = 27 [default = KERNEL_OP_CONV];

  // The scheme used to discretize the convection term.
  optional ConvectionScheme convection_scheme = 2
      [default = CONVECTION_SCHEME_UNKNOWN];
  // The numerical flux for the convection term. Setting default to upwinding
  // for backward compatibility.
  optional NumericalFlux numerical_flux = 35
      [default = NUMERICAL_FLUX_UPWINDING];
  // The scheme used to discretize the diffusion term.
  optional DiffusionScheme diffusion_scheme = 25
      [default = DIFFUSION_SCHEME_CENTRAL_5];
  // The stability limit of diffusion. It sets an upper bound for the viscosity
  // and diffusivity, including those from both kinematic and sub-grid scale
  // (if used), so that nu <= diff_stab_crit h^2 / dt, with h being the minimum
  // of dx, dy, and dz. If it is not set, no upper bound is applied.
  optional float diff_stab_crit = 42;
  // The scheme used for the time integration, which is shared for momentum/
  // velocity and scalars' time advancement by default.
  //
  // When one needs a different time integration scheme for scalars, one could
  // explicitly add `Scalar`'s `time_integration_scheme` field in:
  // swirl_lm/equations/scalars.proto
  // In this case, the scalar specific one will overwrite the shared one here.
  optional TimeIntegrationScheme time_integration_scheme = 3
      [default = TIME_SCHEME_UNKNOWN];

  // The option for how the number of cycles and steps are determined. If not
  // set, `num_cycles` and `num_steps` will be set from flags.
  optional SimulationTimeInfo simulation_time_info = 33;

  // The option for how the pre- and post-process options are specified. If not
  // set, these options will be determined from flags.
  optional PrePostProcessInfo pre_post_process_info = 34;

  // An option for enabling the correction of primitive scalars after the
  // density update from the equation of state. If it is `true`, results of
  // primitive scalars will be updated by the density computed at the current
  // subiteration; otherwise they are stay the same as the ones computed from
  // the scalar-prediction step. Default set to `true` for backward
  // compatibility.
  // Note that scalars oscillations are mitigated with this option being `false`
  // in cases with combustion.
  optional bool enable_scalar_recorrection = 30 [default = true];

  // An option for enabling the Rhie-Chow correction of the face-interpolated
  // momentum in the convection term of the momentum equation. This correction
  // mitigates oscillations in the pressure field and flow field by coupling the
  // two explitictly. Default set to `true` for backward compatibility. Note
  // that this may create instability if the solver is already introducing some
  // degree of coupling between the pressure and the flow field.
  optional bool enable_rhie_chow_correction = 31 [default = true];

  // The boundary conditions for all variables.
  repeated BoundaryCondition boundary_conditions = 5;
  // The number of sub-iterations of the pressure solver.
  optional int32 num_pressure_iterations = 6 [deprecated = true];
  // Options for the pressure solving step.
  optional Pressure pressure = 24;
  // The number of sub-iterations within each time step in the Crank-Nicolson
  // scheme.
  optional int32 num_sub_iterations = 18 [default = 3];
  // The density of the flow field [kg/m^3].
  optional float density = 7 [default = 1.0];
  // The kinematic viscosity of the flow field [m^2/s].
  optional float kinematic_viscosity = 8;
  // The thermodynamic pressure of the flow field [Pa]. Default set to 1 bar,
  // which is the standard atmospheric condition.
  optional float p_thermal = 15 [default = 1.0e5];
  // The procedure for pressure and momentum updates in the solver.
  optional SolverProcedureType solver_procedure = 9 [default = UNKNOWN];
  // The names of transported scalars.
  repeated Scalar scalars = 10;
  // The names of additional states, which are constants that will be used in
  // the simulation, e.g. boundary condition etc.
  repeated string additional_state_keys = 11;
  // The names of the variables that are used as helper in computation. These
  // variables are not subjected to the splitting in z-direction, and they don't
  // have to be of the same size as the mesh.
  repeated string helper_var_keys = 20;
  // The direction vector of gravity if it is considered in the simulation. The
  // default case of (0, 0, 0) corresponds to the case of no gravity.
  optional swirl_lm.CoordinateFloat gravity_direction = 12;
  // The option of using sub-grid scale model. If `False`, the simulation is
  // DNS, otherwise it's an LES. Default solver is DNS.
  optional bool use_sgs = 13 [default = false];
  // The choice of sub-grid scale model to use in the simulation.
  optional SubGridScaleModel sgs_model = 16;
  // The option of updating density. This will be deprecated and replaced by
  // `thermodynamics`.
  optional DensityUpdateOption density_update_option = 14
      [default = DENSITY_UPDATE_UNKNOWN, deprecated = true];
  // The thermodynamic model that's used to update the density.
  optional Thermodynamics thermodynamics = 22;
  // Information required by combustion models.
  optional Combustion combustion = 26;
  // Information required by the turbulent-combustion model. The turbulent-
  // combustion model is a closure model for combustion. It does not provide the
  // reaction source terms due to combustion. It should only be used when
  // a combustion model is defined.
  // Note that the turbulent combustion model is a closure model for combustion
  // in the context of large-eddy simulation. It does not consider the modeling
  // of source terms due to chemical reaction, thus is configured independently
  // from the combustion model.
  optional TurbulentCombustion turbulent_combustion = 38;
  // A filter of variable names to be loaded from files if the filter is used.
  // The option for using this filter will be set by a command line flag.
  repeated string states_from_file = 17;
  // A filter of variable names to be dumped to files.
  repeated string states_to_file = 32;
  // Information required for sponge layers if applied. This will be deprecated,
  // and contained in `boundary_models`.
  optional RayleighDampingLayer sponge_layer = 21 [deprecated = true];
  // Information required for boundary models if applied.
  optional BoundaryModels boundary_models = 23;
  // Specification of analytics monitoring variables.
  optional MonitorSpec monitor_spec = 28;
  // Specification of probes that retrieves values at user specified locations.
  optional Probe probe = 29;
  // If True, all field variables are 3D tf.Tensor. Otherwise, they are
  // Sequence[tf.Tensor].
  optional bool use_3d_tf_tensor = 36 [default = true];
  // The radiation model that solves a two-stream radiative transfer equation
  // taking into account the full atmospheric state.
  optional RadiativeTransfer radiative_transfer = 37;
  // The Lagrangian particle tracking model that solves the motion of particles
  // in a fluid flow.
  optional LagrangianParticleTracking lpt = 41;
  // Names of debug variables to save at the end of each cycle. When adding code
  // to output a new debug variable, update the list below, which serves as
  // documentation for available debug variables.
  //
  // Available Debug Variables
  // -------------------------------------------------------------------------
  //
  repeated string debug_variables = 39;
  // The cloud microphysics parameterization.
  optional Microphysics microphysics = 40;
}
